src/paging.rs

```diff
@@ -506,19 +506,22 @@ impl<T: Translation> PageTableWithLevel<T> {
                 } else {
                     let old = *entry;
                     let (mut subtable, subtable_pa) = Self::new(translation, level + 1);
-                    if let (Some(old_flags), Some(old_pa)) = (old.flags(), old.output_address()) {
-                        // Old was a valid block entry, so we need to split it.
-                        // Recreate the entire block in the newly added table.
-                        let a = align_down(chunk.0.start.0, granularity);
-                        let b = align_up(chunk.0.end.0, granularity);
-                        subtable.map_range(
-                            translation,
-                            &MemoryRegion::new(a, b),
-                            old_pa,
-                            old_flags,
-                        );
+                    if let Some(old_flags) = old.flags() {
+                        if !old_flags.contains(Attributes::TABLE_OR_PAGE) {
+                            let old_pa = old.output_address();
+                            // Old was a valid block entry, so we need to split it.
```

Comment:
How do you know it was valid?

Reply:
Updated, the block no longer needs to have the valid bit.